llama_index/callbacks/async_iter.py

```diff
@@ -0,0 +1,144 @@
+import asyncio
+import logging
+from typing import Any, AsyncIterator, Dict, List, Literal, Optional, Union, cast
+
+from llama_index.callbacks.base_handler import BaseCallbackHandler
+from llama_index.callbacks.schema import CBEventType
+from llama_index.callbacks.utils import superjson_dumps
+
+logger = logging.getLogger(__name__)
+
+
+class AsyncIteratorCallbackHandler(BaseCallbackHandler):
+    """Callback handler that returns an async iterator."""
+
+    queue: asyncio.Queue[str]
+    done: asyncio.Event
+
+    def __init__(
+        self,
+        event_starts_to_ignore: Optional[List[str]] = [],
+        event_ends_to_ignore: Optional[List[str]] = [],
+        verbose: Optional[bool] = True,
+    ) -> None:
+        super().__init__(
+            event_starts_to_ignore=event_starts_to_ignore,
+            event_ends_to_ignore=event_ends_to_ignore,
+        )
+        self.queue = asyncio.Queue()
+        self.done = asyncio.Event()
+        self.verbose = verbose
+
+    def on_event_start(
+        self,
+        event_type: CBEventType,
+        payload: Optional[Dict[str, Any]] = None,
+        event_id: str = "",
+        parent_id: str = "",
+        **kwargs: Any,
+    ) -> str:
+        """Run when an event starts and return id of event."""
+        event = {
+            "event": event_type + ".start",
+            "event_id": event_id,
+            "parent_id": parent_id,
+        }
```

Comment:
Why not put the payload in here as well? That way whatever is reading from the queue will get all the event details.